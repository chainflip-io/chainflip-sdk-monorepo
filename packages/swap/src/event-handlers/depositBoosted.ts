import { arbitrumIngressEgressDepositBoosted as arbitrumSchema160 } from '@chainflip/processor/160/arbitrumIngressEgress/depositBoosted';
import { bitcoinIngressEgressDepositBoosted as bitcoinSchema160 } from '@chainflip/processor/160/bitcoinIngressEgress/depositBoosted';
import { ethereumIngressEgressDepositBoosted as ethereumSchema160 } from '@chainflip/processor/160/ethereumIngressEgress/depositBoosted';
import { polkadotIngressEgressDepositBoosted as polkadotSchema160 } from '@chainflip/processor/160/polkadotIngressEgress/depositBoosted';
import { solanaIngressEgressDepositBoosted as solanaSchema160 } from '@chainflip/processor/160/solanaIngressEgress/depositBoosted';
import { bitcoinIngressEgressDepositBoosted as bitcoinSchema170 } from '@chainflip/processor/170/bitcoinIngressEgress/depositBoosted';
import { arbitrumIngressEgressDepositBoosted as arbitrumSchema180 } from '@chainflip/processor/180/arbitrumIngressEgress/depositBoosted';
import { bitcoinIngressEgressDepositBoosted as bitcoinSchema180 } from '@chainflip/processor/180/bitcoinIngressEgress/depositBoosted';
import { ethereumIngressEgressDepositBoosted as ethereumSchema180 } from '@chainflip/processor/180/ethereumIngressEgress/depositBoosted';
import { polkadotIngressEgressDepositBoosted as polkadotSchema180 } from '@chainflip/processor/180/polkadotIngressEgress/depositBoosted';
import { solanaIngressEgressDepositBoosted as solanaSchema180 } from '@chainflip/processor/180/solanaIngressEgress/depositBoosted';
import { z } from 'zod';
import { assetConstants } from '@/shared/enums';
import { ONE_IN_PIP } from '@/shared/functions';
import { SwapFeeType } from '@/swap/client';
import { getDepositTxRef } from '@/swap/event-handlers/common';
import { EventHandlerArgs } from '.';

const arbitrumSchema = z.union([arbitrumSchema160, arbitrumSchema180]);
const bitcoinSchema = z.union([bitcoinSchema160, bitcoinSchema170, bitcoinSchema180]);
const ethereumSchema = z.union([ethereumSchema160, ethereumSchema180]);
const polkadotSchema = z.union([polkadotSchema160, polkadotSchema180]);
const solanaSchema = z.union([solanaSchema160, solanaSchema180]);

const depositBoostedSchema = z.union([
  arbitrumSchema,
  bitcoinSchema,
  ethereumSchema,
  polkadotSchema,
  solanaSchema,
]);

export type DepositBoostedArgs = z.input<typeof depositBoostedSchema>;

// DepositBoosted event is emitted instead of DepositFinalised event in v140 due to boost
// We need to update the depositAmount and store the ingress fee just like we do in the DepositFinalised event
export const depositBoosted = async ({ prisma, event, block }: EventHandlerArgs) => {
  const {
    asset,
    boostFee,
    action,
    ingressFee,
    amounts,
    prewitnessedDepositId,
    blockHeight,
    ...rest
  } = depositBoostedSchema.parse(event.args);
  const depositDetails = 'depositDetails' in rest ? rest.depositDetails : undefined;
  const maxBoostFeeBps = 'maxBoostFeeBps' in rest ? rest.maxBoostFeeBps : undefined;

  const txRef = getDepositTxRef(assetConstants[asset].chain, depositDetails, blockHeight);

  if (action.__kind === 'Swap' || action.__kind === 'CcmTransfer') {
    const depositAmount = amounts.reduce((acc, [, amount]) => acc + amount, BigInt(0));
    const effectiveBoostFeeBps = (boostFee * BigInt(ONE_IN_PIP)) / BigInt(depositAmount.toString());

    const fees = {
      create: [
        { type: SwapFeeType.BOOST, asset, amount: boostFee.toString() },
        { type: SwapFeeType.INGRESS, asset, amount: ingressFee.toString() },
      ],
    };

    const nativeId = action.swapRequestId;
    const data = {
      maxBoostFeeBps,
      effectiveBoostFeeBps: Number(effectiveBoostFeeBps),
      prewitnessedDepositId,
      depositTransactionRef: txRef,
      depositBoostedAt: new Date(block.timestamp),
      depositBoostedBlockIndex: `${block.height}-${event.indexInBlock}`,
      fees,
    };

    await prisma.swapRequest.update({ where: { nativeId }, data });
  }
};
