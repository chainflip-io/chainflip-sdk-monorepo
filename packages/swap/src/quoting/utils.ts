import assert from 'assert';
import BigNumber from 'bignumber.js';
import { filter, Observable, Subscription } from 'rxjs';
import { Side, type SwapInput } from '@/amm-addon';
import { InternalAsset, getAssetAndChain } from '@/shared/enums';
import { MarketMakerQuote, Leg as MarketMakerLeg } from './schemas';
import env from '../config/env';
import { getAssetPrice } from '../pricing';
import { assertUnreachable } from '../utils/function';
import { SwapRateArgs } from '../utils/statechain';

export type QuoteType = 'pool' | 'market_maker';

export const collectMakerQuotes = (
  requestId: string,
  expectedQuotes: number,
  quotes$: Observable<{ marketMaker: string; quote: MarketMakerQuote }>,
): Promise<MarketMakerQuote[]> => {
  if (expectedQuotes === 0) return Promise.resolve([]);

  const clientsReceivedQuotes = new Map<string, MarketMakerQuote>();

  return new Promise((resolve) => {
    let sub: Subscription;

    const complete = () => {
      sub.unsubscribe();
      resolve([...clientsReceivedQuotes.values()]);
    };

    sub = quotes$
      .pipe(filter(({ quote }) => quote.request_id === requestId))
      .subscribe(({ marketMaker, quote }) => {
        clientsReceivedQuotes.set(marketMaker, quote);
        if (clientsReceivedQuotes.size === expectedQuotes) complete();
      });

    setTimeout(complete, env.QUOTE_TIMEOUT);
  });
};

export const approximateIntermediateOutput = async (asset: InternalAsset, amount: string) => {
  const price = await getAssetPrice(asset);

  if (typeof price !== 'number') return null;

  return BigInt(new BigNumber(amount).times(price).toFixed(0));
};

type BaseAsset = Exclude<InternalAsset, 'Usdc'>;

export class Leg {
  static of(baseAsset: InternalAsset, quoteAsset: InternalAsset, amount: bigint): Leg;
  static of(baseAsset: InternalAsset, quoteAsset: InternalAsset, amount: bigint | null): Leg | null;
  static of(baseAsset: InternalAsset, quoteAsset: InternalAsset, amount: bigint | null) {
    if (amount === null) return null;
    assert(baseAsset !== quoteAsset, 'baseAsset and quoteAsset must be different');
    assert(baseAsset === 'Usdc' || quoteAsset === 'Usdc', 'one of the assets must be USDC');
    return new Leg(baseAsset, quoteAsset, amount);
  }

  private constructor(
    private readonly baseAsset: InternalAsset,
    private readonly quoteAsset: InternalAsset,
    public amount: bigint,
  ) {}

  toPoolJSON(): SwapRateArgs {
    return {
      amount: this.amount,
      srcAsset: this.baseAsset,
      destAsset: this.quoteAsset,
    };
  }

  private getSide(): 'BUY' | 'SELL' {
    return this.quoteAsset !== 'Usdc' ? 'BUY' : 'SELL';
  }

  toMarketMakerJSON(): MarketMakerLeg {
    const side = this.getSide();
    let normalizedBaseAsset: BaseAsset;
    const normalizedQuoteAsset = 'Usdc';

    if (this.quoteAsset !== 'Usdc') {
      normalizedBaseAsset = this.quoteAsset;
    } else if (this.baseAsset !== 'Usdc') {
      normalizedBaseAsset = this.baseAsset;
    } else {
      return assertUnreachable('invalid leg');
    }

    return {
      base_asset: getAssetAndChain(normalizedBaseAsset),
      quote_asset: getAssetAndChain(normalizedQuoteAsset),
      amount: this.amount.toString(),
      side,
    };
  }

  toSwapInput(): SwapInput {
    return {
      side: this.getSide() === 'BUY' ? Side.Buy : Side.Sell,
      amount: this.amount,
      limitOrders: [],
    };
  }
}
