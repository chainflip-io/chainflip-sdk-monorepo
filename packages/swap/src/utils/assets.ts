import { SupportedAsset, supportedAsset } from '@/shared/enums';

export const isSupportedAsset = (value: string): value is SupportedAsset =>
  supportedAsset.safeParse(value).success;

export function assertSupportedAsset(
  value: string,
): asserts value is SupportedAsset {
  if (!isSupportedAsset(value)) {
    throw new Error(`received invalid asset "${value}"`);
  }
}

export const decimalPlaces: Record<SupportedAsset, number> = {
  DOT: 10,
  ETH: 18,
  FLIP: 18,
  USDC: 6,
  BTC: 8,
};
