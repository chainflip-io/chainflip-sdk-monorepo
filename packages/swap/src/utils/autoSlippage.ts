import BigNumber from 'bignumber.js';
import { assetConstants, chainConstants } from '@/shared/enums';
import { getDeployedLiquidity, getUndeployedLiquidity } from './pools';
import { getRequiredBlockConfirmations } from './rpc';
import { InternalAsset } from '../client';

const getLiquidityAutoSlippage = async (
  srcAsset: InternalAsset,
  destAsset: InternalAsset,
  amount: bigint,
) => {
  const deployedLiquidity = await getDeployedLiquidity(srcAsset, destAsset);
  const liquidityRatio = new BigNumber(amount.toString()).div(deployedLiquidity.toString());

  if (liquidityRatio.isGreaterThan(0.2)) {
    return liquidityRatio.multipliedBy(2);
  }
  return new BigNumber(0);
};

const getDepositTimeAutoSlippage = async (srcAsset: InternalAsset, isBoosted: boolean) => {
  const { chain } = assetConstants[srcAsset];
  const { blockTimeSeconds } = chainConstants[chain];
  const blockConfirmations = isBoosted ? 1 : (await getRequiredBlockConfirmations(srcAsset)) ?? 0;
  const depositTimeMinutes = (blockConfirmations * blockTimeSeconds) / 60;

  if (['Solana', 'Arbitrum'].includes(chain)) {
    return depositTimeMinutes * 0.01;
  }
  if (['Ethereum', 'Polkadot'].includes(chain)) {
    return depositTimeMinutes * 0.05;
  }
  return depositTimeMinutes * 0.1;
};

const getUndeployedLiquidityAutoSlippage = async (asset: InternalAsset, amount: bigint) => {
  const undeployedLiquidity = await getUndeployedLiquidity(asset);
  if (undeployedLiquidity >= amount) {
    return new BigNumber(-0.5);
  }

  const undeployedRatio = new BigNumber(undeployedLiquidity.toString()).div(amount.toString());
  if (undeployedRatio.gt(0.5)) {
    return new BigNumber(-undeployedRatio * 0.1);
  }

  return new BigNumber(0);
};

export const calculateRecommendedSlippage = async ({
  srcAsset,
  destAsset,
  intermediateAmount,
  egressAmount,
  boostFeeBps,
  dcaChunks,
}: {
  srcAsset: InternalAsset;
  destAsset: InternalAsset;
  intermediateAmount?: bigint;
  egressAmount: bigint;
  boostFeeBps?: number;
  dcaChunks: number;
}) => {
  const minSlippage = 0.1;
  const baseSlippage = 1;

  const assetTo = srcAsset === 'Usdc' || destAsset === 'Usdc' ? destAsset : 'Usdc';

  const [deployedLiquiditySlippage, timeSlippage, undeployedLiquiditySlippage] = await Promise.all([
    await getLiquidityAutoSlippage(
      srcAsset,
      assetTo,
      (intermediateAmount ?? egressAmount) * BigInt(dcaChunks),
    ),
    await getDepositTimeAutoSlippage(srcAsset, Boolean(boostFeeBps && boostFeeBps > 0)),
    await getUndeployedLiquidityAutoSlippage(
      intermediateAmount ? 'Usdc' : destAsset,
      (intermediateAmount ?? egressAmount) * BigInt(dcaChunks),
    ),
  ]);

  const calculatedSlippage = deployedLiquiditySlippage
    .plus(timeSlippage)
    .plus(undeployedLiquiditySlippage)
    .plus(baseSlippage);

  return Math.max(minSlippage, Math.round(calculatedSlippage.toNumber() * 100) / 100);
};
