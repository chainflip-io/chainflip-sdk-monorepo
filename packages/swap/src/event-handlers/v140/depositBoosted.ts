import { arbitrumIngressEgressDepositBoosted } from '@chainflip/processor/160/arbitrumIngressEgress/depositBoosted';
import { bitcoinIngressEgressDepositBoosted } from '@chainflip/processor/160/bitcoinIngressEgress/depositBoosted';
import { ethereumIngressEgressDepositBoosted } from '@chainflip/processor/160/ethereumIngressEgress/depositBoosted';
import { polkadotIngressEgressDepositBoosted } from '@chainflip/processor/160/polkadotIngressEgress/depositBoosted';
import { solanaIngressEgressDepositBoosted } from '@chainflip/processor/160/solanaIngressEgress/depositBoosted';
import { z } from 'zod';
import { ONE_IN_PIP } from '@/shared/functions';
import { SwapFeeType } from '@/swap/client';
import { EventHandlerArgs } from '..';

const arbitrumSchema = arbitrumIngressEgressDepositBoosted;
const bitcoinSchema = bitcoinIngressEgressDepositBoosted;
const ethereumSchema = ethereumIngressEgressDepositBoosted;
const polkadotSchema = polkadotIngressEgressDepositBoosted;
const solanaSchema = solanaIngressEgressDepositBoosted;

const depositBoostedSchema = z.union([
  arbitrumSchema,
  bitcoinSchema,
  ethereumSchema,
  polkadotSchema,
  solanaSchema,
]);

export type DepositBoostedArgs = z.input<typeof depositBoostedSchema>;

// DepositBoosted event is emitted instead of DepositFinalised event in v140 due to boost
// We need to update the depositAmount and store the ingress fee just like we do in the DepositFinalised event
export const depositBoosted = async ({ prisma, event, block }: EventHandlerArgs) => {
  const { asset, boostFee, action, ingressFee, amounts, prewitnessedDepositId } =
    depositBoostedSchema.parse(event.args);

  if (action.__kind === 'Swap') {
    const depositAmount = amounts.reduce((acc, [, amount]) => acc + amount, BigInt(0));
    const effectiveBoostFeeBps = (boostFee * BigInt(ONE_IN_PIP)) / BigInt(depositAmount.toString());

    const fees = {
      create: [
        { type: SwapFeeType.BOOST, asset, amount: boostFee.toString() },
        { type: SwapFeeType.INGRESS, asset, amount: ingressFee.toString() },
      ],
    };

    const nativeId = action.swapRequestId;
    const data = {
      effectiveBoostFeeBps: Number(effectiveBoostFeeBps),
      prewitnessedDepositId,
      depositBoostedAt: new Date(block.timestamp),
      depositBoostedBlockIndex: `${block.height}-${event.indexInBlock}`,
      fees,
    };

    await prisma.swapRequest.update({ where: { nativeId }, data });
  }
};
