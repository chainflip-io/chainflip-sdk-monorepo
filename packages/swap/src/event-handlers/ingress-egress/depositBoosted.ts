import { arbitrumIngressEgressDepositBoosted as arbitrumSchema180 } from '@chainflip/processor/180/arbitrumIngressEgress/depositBoosted';
import { bitcoinIngressEgressDepositBoosted as bitcoinSchema180 } from '@chainflip/processor/180/bitcoinIngressEgress/depositBoosted';
import { ethereumIngressEgressDepositBoosted as ethereumSchema180 } from '@chainflip/processor/180/ethereumIngressEgress/depositBoosted';
import { polkadotIngressEgressDepositBoosted as polkadotSchema180 } from '@chainflip/processor/180/polkadotIngressEgress/depositBoosted';
import { solanaIngressEgressDepositBoosted as solanaSchema180 } from '@chainflip/processor/180/solanaIngressEgress/depositBoosted';
import { arbitrumIngressEgressDepositBoosted as arbitrumSchema190 } from '@chainflip/processor/190/arbitrumIngressEgress/depositBoosted';
import { assethubIngressEgressDepositBoosted as assethubSchema190 } from '@chainflip/processor/190/assethubIngressEgress/depositBoosted';
import { bitcoinIngressEgressDepositBoosted as bitcoinSchema190 } from '@chainflip/processor/190/bitcoinIngressEgress/depositBoosted';
import { ethereumIngressEgressDepositBoosted as ethereumSchema190 } from '@chainflip/processor/190/ethereumIngressEgress/depositBoosted';
import { polkadotIngressEgressDepositBoosted as polkadotSchema190 } from '@chainflip/processor/190/polkadotIngressEgress/depositBoosted';
import { solanaIngressEgressDepositBoosted as solanaSchema190 } from '@chainflip/processor/190/solanaIngressEgress/depositBoosted';
import { ChainflipChain } from '@chainflip/utils/chainflip';
import { z } from 'zod';
import { ONE_IN_PIP } from '@/shared/functions.js';
import { Prisma, SwapFeeType } from '../../client.js';
import { getDepositTxRef } from '../common.js';
import { EventHandlerArgs } from '../index.js';

const arbitrumSchema = z.union([arbitrumSchema190, arbitrumSchema180]).transform((args) => ({
  ...args,
  depositDetails: { chain: 'Arbitrum' as const, data: args.depositDetails },
}));
const bitcoinSchema = z.union([bitcoinSchema190, bitcoinSchema180]).transform((args) => ({
  ...args,
  depositDetails: { chain: 'Bitcoin' as const, data: args.depositDetails },
}));
const ethereumSchema = z.union([ethereumSchema190, ethereumSchema180]).transform((args) => ({
  ...args,
  depositDetails: { chain: 'Ethereum' as const, data: args.depositDetails },
}));
const polkadotSchema = z.union([polkadotSchema190, polkadotSchema180]).transform((args) => ({
  ...args,
  depositDetails: { chain: 'Polkadot' as const, data: args.depositDetails },
}));
const solanaSchema = z.union([solanaSchema190, solanaSchema180]).transform((args) => ({
  ...args,
  depositDetails: { chain: 'Solana' as const, data: undefined },
}));
const assethubSchema = assethubSchema190.transform((args) => ({
  ...args,
  depositDetails: { chain: 'Assethub' as const, data: args.depositDetails },
}));

const schemas = {
  Arbitrum: arbitrumSchema,
  Bitcoin: bitcoinSchema,
  Ethereum: ethereumSchema,
  Polkadot: polkadotSchema,
  Solana: solanaSchema,
  Assethub: assethubSchema,
} as const satisfies Record<ChainflipChain, z.ZodTypeAny>;

export type DepositBoostedArgsMap = {
  [C in ChainflipChain]: z.input<(typeof schemas)[C]>;
};

// DepositBoosted event is emitted instead of DepositFinalised event in v140 due to boost
// We need to update the depositAmount and store the ingress fee just like we do in the DepositFinalised event
export const depositBoosted =
  (chain: ChainflipChain) =>
  async ({ prisma, event, block }: EventHandlerArgs) => {
    const {
      asset,
      boostFee,
      action,
      ingressFee,
      amounts,
      prewitnessedDepositId,
      blockHeight,
      depositDetails,
      maxBoostFeeBps,
    } = schemas[chain].parse(event.args);

    const txRef = getDepositTxRef(depositDetails, blockHeight);

    if (action.__kind === 'Swap' || action.__kind === 'CcmTransfer') {
      const depositAmount = amounts.reduce((acc, [, amount]) => acc + amount, BigInt(0));
      const effectiveBoostFeeBps =
        (boostFee * BigInt(ONE_IN_PIP)) / BigInt(depositAmount.toString());

      const fees = {
        create: [
          { type: SwapFeeType.BOOST, asset, amount: boostFee.toString() },
          { type: SwapFeeType.INGRESS, asset, amount: ingressFee.toString() },
        ],
      };

      const nativeId = action.swapRequestId;
      const data: Prisma.SwapRequestUpdateInput = {
        maxBoostFeeBps,
        effectiveBoostFeeBps: Number(effectiveBoostFeeBps),
        prewitnessedDepositId,
        depositTransactionRef: txRef,
        depositBoostedAt: new Date(block.timestamp),
        depositBoostedBlockIndex: `${block.height}-${event.indexInBlock}`,
        fees,
        depositAmount: depositAmount.toString(),
      };

      await prisma.swapRequest.update({ where: { nativeId }, data });
    }
  };
