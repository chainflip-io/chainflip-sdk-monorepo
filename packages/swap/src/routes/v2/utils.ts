import { getAssetAndChain } from '@/shared/enums';
import { assertUnreachable } from '@/shared/functions';
import { StateV2 } from './swap';
import prisma, {
  Prisma,
  Swap,
  Egress,
  Broadcast,
  SwapFee,
  IgnoredEgress,
  FailedSwap,
} from '../../client';
import { getPendingBroadcast } from '../../ingress-egress-tracking';
import { failedSwapMessage } from '../../utils/swap';
import { FailureMode } from '../swap';

export const depositChannelInclude = {
  failedBoosts: true,
  failedSwaps: true,
  affiliates: {
    select: {
      account: true,
      commissionBps: true,
    },
  },
} as const;

export const swapRequestInclude = {
  swaps: { include: { fees: true }, orderBy: { nativeId: 'desc' } },
  egress: { include: { broadcast: true } },
  refundEgress: { include: { broadcast: true } },
  fees: true,
  ignoredEgresses: true,
  swapDepositChannel: {
    include: depositChannelInclude,
  },
} as const;

export const getSwapFields = (swap: Swap & { fees: SwapFee[] }) => ({
  swapInputAmount: swap.swapInputAmount.toString(),
  swapOutputAmount: swap.swapOutputAmount?.toString(),
  scheduledAt: swap.swapScheduledAt.valueOf(),
  scheduledBlockIndex: swap.swapScheduledBlockIndex ?? undefined,
  executedAt: swap.swapExecutedAt?.valueOf(),
  executedBlockIndex: swap.swapExecutedBlockIndex ?? undefined,
  retryCount: swap.retryCount,
  latestSwapRescheduledAt: swap.latestSwapRescheduledAt?.valueOf(),
  latestSwapRescheduledBlockIndex: swap.latestSwapRescheduledBlockIndex ?? undefined,
  fees: swap.fees.map((fee) => ({
    type: fee.type,
    ...getAssetAndChain(fee.asset),
    amount: fee.amount.toString(),
  })),
});

export const getDepositIgnoredFailedState = (failedSwap: FailedSwap) => ({
  failedAt: failedSwap.failedAt.valueOf(),
  failedAtBlockIndex: failedSwap.failedBlockIndex,
  mode: FailureMode.IngressIgnored,
  reason: {
    name: failedSwap.reason,
    message: failedSwapMessage[failedSwap.reason],
  },
});

export const getEgressFailureState = async (
  ignoredEgress: IgnoredEgress | undefined,
  broadcast: Broadcast | null | undefined,
  type: IgnoredEgress['type'] | undefined,
) => {
  const abortedBroadcast = broadcast?.abortedAt?.valueOf();
  const failedAt = abortedBroadcast ?? ignoredEgress?.ignoredAt?.valueOf();
  const failedAtBlockIndex = broadcast?.abortedBlockIndex ?? ignoredEgress?.ignoredBlockIndex;

  let error;
  let failureMode;

  if (failedAt && failedAtBlockIndex) {
    error = {
      name: 'Unknown',
      message: 'An unknown error occurred',
    };
    if (ignoredEgress) {
      switch (ignoredEgress.type) {
        case 'REFUND':
          failureMode = FailureMode.RefundEgressIgnored;
          break;
        case 'SWAP':
          failureMode = FailureMode.EgressIgnored;
          break;
        default:
          assertUnreachable(ignoredEgress.type);
      }
      const [stateChainError] = await Promise.all([
        prisma.stateChainError.findUniqueOrThrow({
          where: { id: ignoredEgress.stateChainErrorId },
        }),
      ]);
      error = {
        name: stateChainError.name,
        message: stateChainError.docs,
      };
    } else if (abortedBroadcast) {
      const message =
        type === 'SWAP' ? 'The swap broadcast was aborted' : 'The refund broadcast was aborted';
      failureMode = FailureMode.BroadcastAborted;
      error = {
        name: 'BroadcastAborted',
        message,
      };
    }
    return {
      failedAt,
      failedAtBlockIndex,
      mode: failureMode,
      reason: error,
    };
  }
  return undefined;
};

export const getEgressStatusFields = async (
  egress: Egress | null | undefined,
  broadcast: Broadcast | null | undefined,
  ignoredEgresses: IgnoredEgress[] | undefined,
  type: IgnoredEgress['type'] | undefined,
  egressTrackerTxRef: string | null | undefined,
) => {
  const ignoredEgress = ignoredEgresses?.find((e) => e.type === type);
  const failureState = await getEgressFailureState(ignoredEgress, broadcast, type);
  return {
    ...(egress && {
      outputAmount: egress.amount?.toFixed(),
      scheduledAt: egress.scheduledAt?.valueOf(),
      scheduledBlockIndex: egress.scheduledBlockIndex ?? undefined,
    }),
    ...(broadcast && {
      sentAt: broadcast?.succeededAt?.valueOf(),
      sentAtBlockIndex: broadcast?.succeededBlockIndex ?? undefined,
      sentTxRef: broadcast?.transactionRef ?? egressTrackerTxRef,
    }),
    ...(failureState && { failure: failureState }),
    ...(ignoredEgress && { ignoredAmount: ignoredEgress.amount?.toFixed() }),
  };
};

export const getSwapState = async (
  failedSwap: FailedSwap | null | undefined,
  ignoredEgresses: IgnoredEgress[] | undefined,
  swapRequest:
    | Prisma.SwapRequestGetPayload<{
        include: typeof swapRequestInclude;
      }>
    | undefined
    | null,
) => {
  let state: StateV2 | undefined;
  let swapEgressTrackerTxRef: string | null | undefined;
  let refundEgressTrackerTxRef: string | null | undefined;
  const swapEgress = swapRequest?.egress;
  const refundEgress = swapRequest?.refundEgress;
  const egress = swapEgress ?? refundEgress;

  if (
    failedSwap ||
    (ignoredEgresses && ignoredEgresses.length > 0) ||
    egress?.broadcast?.abortedAt
  ) {
    state = StateV2.Failed;
  } else if (egress?.broadcast?.succeededAt) {
    state = StateV2.Completed;
  } else if (egress?.broadcast) {
    if (swapEgress?.broadcast) {
      state = StateV2.Sending;
      const pendingSwapBroadcast = await getPendingBroadcast(swapEgress.broadcast);
      if (pendingSwapBroadcast) {
        state = StateV2.Sent;
        swapEgressTrackerTxRef = pendingSwapBroadcast.tx_ref;
      }
    }
    if (refundEgress?.broadcast) {
      const pendingRefundBroadcast = await getPendingBroadcast(refundEgress.broadcast);
      if (pendingRefundBroadcast) {
        state = StateV2.Sent;
        refundEgressTrackerTxRef = pendingRefundBroadcast.tx_ref;
      }
    }
  } else if (egress) {
    state = StateV2.Sending;
  } else if (swapRequest?.swaps.some((s) => s.swapScheduledAt)) {
    state = StateV2.Swapping;
  } else {
    state = StateV2.Receiving;
  }

  return {
    state,
    swapEgressTrackerTxRef,
    refundEgressTrackerTxRef,
  };
};
