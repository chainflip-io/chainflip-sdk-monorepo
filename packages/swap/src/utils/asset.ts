import { ChainflipAsset } from '@chainflip/utils/chainflip';

export const isAssetHub = (asset: ChainflipAsset): asset is 'HubDot' | 'HubUsdc' | 'HubUsdt' =>
  asset === 'HubDot' || asset === 'HubUsdc' || asset === 'HubUsdt';
