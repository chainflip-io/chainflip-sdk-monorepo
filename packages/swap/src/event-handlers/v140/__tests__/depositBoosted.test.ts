import { z } from 'zod';
import { actionSchema } from '@/shared/parsers';
import { createBtcSwapDepositChannel } from './utils';
import prisma from '../../../client';
import { depositBoosted } from '../depositBoosted';

export const depositBoostedBtcMock = ({
  action,
  amounts,
  channelId,
}: {
  action?: z.input<typeof actionSchema>;
  amounts?: [[number, string]];
  channelId?: string;
}) =>
  ({
    block: {
      height: 120,
      timestamp: 1670337105000,
    },
    eventContext: {
      kind: 'event',
      event: {
        args: {
          asset: {
            __kind: 'Btc',
          },
          amounts: amounts ?? [[5, '1000000']],
          prewitnessedDepositId: '101',
          channelId: channelId ?? '1',
          ingressFee: '1000',
          boostFee: '500',
          action: action ?? { __kind: 'Swap', swapId: '1' },
        },
        name: 'BitcoinIngressEgress.DepositBoosted',
        indexInBlock: 7,
      },
    },
  }) as const;

describe('depositBoosted', () => {
  beforeEach(async () => {
    await prisma.$queryRaw`TRUNCATE TABLE "private"."DepositChannel" CASCADE`;
    await prisma.$queryRaw`TRUNCATE TABLE "SwapDepositChannel", "Swap" CASCADE`;
  });

  it('updates the values for an existing swap', async () => {
    const swapDepositChannel = await createBtcSwapDepositChannel();
    const eventData = depositBoostedBtcMock({ amounts: [[5, '1000000']] }) as any;
    const event = eventData.eventContext.event as any;
    const block = eventData.block as any;

    await prisma.$transaction(async (txClient) => {
      await depositBoosted({
        prisma: txClient,
        event,
        block,
      });
    });

    const request = await prisma.swapRequest.findFirstOrThrow({
      where: { swapDepositChannelId: swapDepositChannel.id },
      include: { fees: true },
    });

    expect(request.depositAmount.toString()).toBe('1000000');
    expect(request.effectiveBoostFeeBps).toBe(5);

    expect(request).toMatchSnapshot({
      id: expect.any(BigInt),
      swapDepositChannelId: expect.any(BigInt),
      depositBoostedAt: expect.any(Date),
      depositBoostedBlockIndex: expect.any(String),
      fees: [
        { id: expect.any(BigInt), swapId: expect.any(BigInt), type: 'BOOST' },
        { id: expect.any(BigInt), swapId: expect.any(BigInt), type: 'INGRESS' },
      ],
    });
  });

  it('does nothing when the action is not a swap', async () => {
    const swapDepositChannel = await createBtcSwapDepositChannel({});
    const eventData = depositBoostedBtcMock({
      amounts: [[5, '1000000']],
      action: { __kind: 'LiquidityProvision', lpAccount: '0x123' },
    }) as any;
    const event = eventData.eventContext.event as any;
    const block = eventData.block as any;

    await prisma.$transaction(async (txClient) => {
      await depositBoosted({
        prisma: txClient,
        event,
        block,
      });
    });

    const request = await prisma.swapRequest.findFirstOrThrow({
      where: { swapDepositChannelId: swapDepositChannel.id },
      include: { fees: true },
    });

    expect(request.effectiveBoostFeeBps).toBeNull();

    expect(request).toMatchSnapshot({
      id: expect.any(BigInt),
      swapDepositChannelId: expect.any(BigInt),
      fees: [],
    });
  });
});
