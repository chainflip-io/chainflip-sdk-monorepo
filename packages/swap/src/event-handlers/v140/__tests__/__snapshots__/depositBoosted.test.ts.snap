// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`depositBoosted does nothing when the action is not a swap 1`] = `
{
  "ccmDepositReceivedBlockIndex": null,
  "ccmGasBudget": null,
  "ccmMessage": null,
  "createdAt": Any<Date>,
  "depositAmount": "1000000",
  "depositBoostedAt": null,
  "depositBoostedBlockIndex": null,
  "depositReceivedAt": 2022-12-06T14:31:39.000Z,
  "depositReceivedBlockIndex": "0-15",
  "destAddress": "0x6fd76a7699e6269af49e9c63f01f61464ab21d1c",
  "destAsset": "Eth",
  "effectiveBoostFeeBps": null,
  "egressId": null,
  "fees": [],
  "id": Any<BigInt>,
  "intermediateAmount": null,
  "nativeId": 1n,
  "srcAsset": "Btc",
  "swapDepositChannelId": Any<BigInt>,
  "swapExecutedAt": null,
  "swapExecutedBlockIndex": null,
  "swapInputAmount": "1000000",
  "swapOutputAmount": null,
  "txHash": null,
  "type": "SWAP",
  "updatedAt": Any<Date>,
}
`;

exports[`depositBoosted updates the values for an existing swap 1`] = `
{
  "ccmDepositReceivedBlockIndex": null,
  "ccmGasBudget": null,
  "ccmMessage": null,
  "createdAt": Any<Date>,
  "depositAmount": "1000000",
  "depositBoostedAt": Any<Date>,
  "depositBoostedBlockIndex": Any<String>,
  "depositReceivedAt": 2022-12-06T14:31:39.000Z,
  "depositReceivedBlockIndex": "0-15",
  "destAddress": "0x6fd76a7699e6269af49e9c63f01f61464ab21d1c",
  "destAsset": "Eth",
  "effectiveBoostFeeBps": 5,
  "egressId": null,
  "fees": [
    {
      "amount": "500",
      "asset": "Btc",
      "id": Any<BigInt>,
      "swapId": Any<BigInt>,
      "type": "BOOST",
    },
    {
      "amount": "1000",
      "asset": "Btc",
      "id": Any<BigInt>,
      "swapId": Any<BigInt>,
      "type": "INGRESS",
    },
  ],
  "id": Any<BigInt>,
  "intermediateAmount": null,
  "nativeId": 1n,
  "srcAsset": "Btc",
  "swapDepositChannelId": Any<BigInt>,
  "swapExecutedAt": null,
  "swapExecutedBlockIndex": null,
  "swapInputAmount": "1000000",
  "swapOutputAmount": null,
  "txHash": null,
  "type": "SWAP",
  "updatedAt": Any<Date>,
}
`;
